File: typescript/leetcodehard/longestIncreasingPathMatrix.ts
/*
https://leetcode.com/problems/longest-increasing-path-in-a-matrix/

status - completed
*/

import { logp } from "../utils/logger";

function longestIncreasingPath(matrix: number[][]): number {
  const rows = matrix.length;
  const cols = matrix[0].length;
  const memo: number[][] = Array.from({ length: rows }, () => new Array(cols).fill(0));
  const directions = [
    [1, 0],
    [-1, 0],
    [0, 1],
    [0, -1],
  ];

  function dfs(row: number, col: number): number {
    if (memo[row][col] !== 0) return memo[row][col];

    let best = 1;
    for (const [dr, dc] of directions) {
      const nr = row + dr;
      const nc = col + dc;
      if (nr < 0 || nr >= rows || nc < 0 || nc >= cols) continue;
      if (matrix[nr][nc] <= matrix[row][col]) continue;
      best = Math.max(best, 1 + dfs(nr, nc));
    }

    memo[row][col] = best;
    return best;
  }

  let result = 0;
  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      result = Math.max(result, dfs(row, col));
    }
  }

  return result;
}

logp(
  `input [[9,9,4],[6,6,8],[2,1,1]] expected 4 actual ${longestIncreasingPath([
    [9, 9, 4],
    [6, 6, 8],
    [2, 1, 1],
  ])}`,
);
logp(
  `input [[3,4,5],[3,2,6],[2,2,1]] expected 4 actual ${longestIncreasingPath([
    [3, 4, 5],
    [3, 2, 6],
    [2, 2, 1],
  ])}`,
);

File: python/leetcodehard/longest_increasing_path_matrix.py
"""
https://leetcode.com/problems/longest-increasing-path-in-a-matrix/

status - completed
"""

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def longestIncreasingPath(self, matrix):
        if not matrix or not matrix[0]:
            return 0

        rows = len(matrix)
        cols = len(matrix[0])
        memo = [[0] * cols for _ in range(rows)]

        def dfs(r, c):
            if memo[r][c]:
                return memo[r][c]

            best = 1
            for dr, dc in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0 <= nc < cols and matrix[nr][nc] > matrix[r][c]:
                    best = max(best, 1 + dfs(nr, nc))

            memo[r][c] = best
            return best

        longest = 0
        for r in range(rows):
            for c in range(cols):
                longest = max(longest, dfs(r, c))

        return longest


if __name__ == "__main__":
    test_input_1 = [[9, 9, 4], [6, 6, 8], [2, 1, 1]]
    output = Solution().longestIncreasingPath(test_input_1)
    Solution().log.info("Output is %s", output)

    test_input_2 = [[3, 4, 5], [3, 2, 6], [2, 2, 1]]
    output = Solution().longestIncreasingPath(test_input_2)
    Solution().log.info("Output is %s", output)
